README.md
# Atlassian Jira for Docker

__docker-atlassian-jira__ provides [Atlassian Jira](https://www.atlassian.com/software/jira) in a [docker](https://www.docker.com/)
container to support issue and project tracking for software teams.

## Overview

This revision of __docker-atlassian-jira__ will install:

[JIRA 7.7.9](https://confluence.atlassian.com/jirasoftware/issues-resolved-in-7-7-0-942847220.html)

## Installation

This application is ready to launch on a Docker host:

```sh
prompt> docker run -d -p 8080:8080 -p 8443:8443 markeissler/atlassian-jira:latest
```

## Usage

<a name="data-persistence"></a>

### Data Persistence

As configured, data on the following volumes will be created to persist data between container starts:

| Volume | Purpose                                                    |
|:-------|:-----------------------------------------------------------|
| /var/atlassian/jira                     | application configuration |
| /opt/atlassian/jira/logs                | runtime logs              |

<a name="data-persistence-nfs"></a>

### Data Persistence over NFS

It may be desirable to configure data persistence over NFS, in which case NFS volumes are mounted at the locations
described in the [Data Persistence](#data-persistence) section above. NFS support requires that the underlying Docker
host supports NFS; if deploying to a [Docker swarm](https://docs.docker.com/engine/swarm/) a potential __boot2docker.iso__
candidate that supports NFS is the [boot2docker-nfs.iso](https://github.com/markeissler/boot2docker-nfs).

Certain JIRA directories are moved out of the application configuration directory and into an ephemeral runtime storage
area to prevent data corruption startup failures. Specfically, cache directories are moved so that clean re-starts
are possible; often, when an instance dies Tomcat will not be shutdown cleanly and data corruption is likely to occur
with regard to the _felix_ plugin cache).

| Directory | Purpose                                                    |
|:----------|:-----------------------------------------------------------|
| /var/atlassian/jira_runtime   | runtime storage for caches and indexes |

### SSL Support

You can enable SSL by simply copying a PKCS12 format certificate (`certificate.p12`) into the `JIRA_HOME` directory
(`/var/atlassian/jira`) and then restarting the container. The PKCS12 file format has been selected to make it easier to
generate certificates using `openssl`.

An example `openssl` command that will create a PKCS12 file from a private key (`server_key.pem`) and public certficate
(`server_cert.pem`) follows:

```sh
prompt> openssl pkcs12 -export -in server_cert.pem \
    -inkey server_key.pem -out certificate.p12 \
    -passout pass:changeit -name "jira"
```

On container startup, the PCKS12 format certificate.p12 file will be converted and stored in the system JKS keystore.

## Docker Swarm Support

While __docker-atlassian-jira__ does not support multi-node clustering it does support deployment to a cluster with a
failover configuration (where only a single JIRA instance is active at any time).

This configuration requires that [Data Persistence over NFS](#data-persistence-nfs) has been configured to share JIRA
configuration information among replicated instances.

## Redeployment Index Regeneration

Indexes are stored in ephemeral storage; consequently, when redeploying the application on top of existing persistent
storage (i.e. [Data Persistence over NFS](#data-persistence-nfs)) it will be necessary to manually trigger a rebuild of
all indexes. Log in as a user with administrative permissions, then navigate to:

`Cog > System > Advanced > Indexing`

Click on the <kbd>Re-Index</kbd> button.

## Troubleshooting

For general troubleshooting information check the [Troubleshoot](troubleshoot.md) document.

## Upstream Release Notes

Release notes for all versions of JIRA can be viewed online:

[JIRA Release Notes](https://confluence.atlassian.com/jirasoftware/jira-software-release-notes-776821069.html)

## Authors

__docker-atlassian-jira__ is the work of __Mark Eissler__.

## Attributions

__docker-atlassian-jira__ was inspired by the work of [Martin Aksel Jensen](https://github.com/cptactionhank),
specifically his ongoing efforts to provide up-to-date _dockerized_ versions of other popular [Atlassian](https://www.atlassian.com/)
applications.

## License

__docker-atlassian-jira__ is licensed under the MIT open source license.

---
Without open source, there would be no Internet as we know it today.


Changelog.md
# Changelog: docker-atlassian-jira

## 1.15.0 / 2018-01-28

Update JIRA to 7.7.0

### Minor version increment!

Since this is a Minor version increment be sure to review the release notes and upgrade notes:

  * [JIRA Software 7.7.x release notes](https://confluence.atlassian.com/jirasoftware/jira-software-7-7-x-release-notes-941594462.html)
  * [JIRA Software 7.7.x upgrade notes](https://confluence.atlassian.com/jirasoftware/jira-software-7-7-x-upgrade-notes-941594485.html)

## 1.14.0 / 2018-01-28

Update JIRA to 7.6.3

## 1.13.0 / 2018-01-28

Update JIRA to 7.6.2

## 1.12.0 / 2018-01-28

Update JIRA to 7.6.1

## 1.11.0 / 2018-01-28

Update JIRA to 7.6.0

### Minor version increment!

Since this is a Minor version increment be sure to review the release notes and upgrade notes:

  * [JIRA Software 7.6.x release notes](https://confluence.atlassian.com/jirasoftware/jira-software-7-6-x-release-notes-938027181.html)
  * [JIRA Software 7.6.x upgrade notes](https://confluence.atlassian.com/jirasoftware/jira-software-7-6-x-upgrade-notes-938027185.html)

## 1.10.0 / 2018-01-28

Update JIRA to 7.5.4

## 1.9.0 / 2017-12-31

Update JIRA to 7.5.3

## 1.8.0 / 2017-11-05

Update JIRA to 7.5.2.

## 1.7.0 / 2017-11-05

Update JIRA to 7.5.1.

## 1.6.0 / 2017-09-13

Update JIRA to 7.5.0.

### Minor version increment!

Since this is a Minor version increment be sure to review the release notes and upgrade notes:

  * [JIRA Software 7.5.x release notes](https://confluence.atlassian.com/jirasoftware/jira-software-7-5-x-release-notes-934719297.html)
  * [JIRA Software 7.5.x upgrade notes](https://confluence.atlassian.com/jirasoftware/jira-software-7-5-x-upgrade-notes-934719299.html)

## 1.5.0 / 2017-09-13

Update JIRA to 7.4.4.

## 1.4.0 / 2017-09-13

Update JIRA to 7.4.3.

## 1.3.0 / 2017-09-13

Update JIRA to 7.4.2.

## 1.2.0 / 2017-07/27

Update JIRA to 7.4.1.

### Short list of commit messages

  * Update JIRA to 7.4.1.

## 1.1.0 / 2017-07-10

Update JIRA to 7.4.0.

### Minor version increment!

Since this is a Minor version increment be sure to review the release notes and upgrade notes:

  * [JIRA Software 7.4.x release notes](https://confluence.atlassian.com/jirasoftware/jira-software-7-4-x-release-notes-902079634.html)
  * [JIRA Software 7.4.x upgrade notes](https://confluence.atlassian.com/jirasoftware/jira-software-7-4-x-upgrade-notes-907283484.html)

### Short list of commit messages

  * Update JIRA to 7.4.0.

## 1.0.0 / 2017-06-25

Update JIRA to 7.3.8.

### Short list of commit messages

  * Update README to include link to upstream release notes.
  * Update README for redeployment indexing issues.
  * Update JIRA to 7.3.8.

## 0.11.0 / 2017-06-19

Docker Swarm support! This version adds support for deployment to a cluster with a failover configuration. That is, only
one instance can be active at a time but the failover instance should startup without encountering errors stemming from
a corrupted _felix_ plugin cache.

Lucene indexes have been moved out of the Docker host's persistence directory so that configurations that enable data
persistence with NFS can be better supported (Lucene doesn't get along with NFS).

### Short list of commit messages

  * Update README for ephemeral storage and Swarm support.
  * Use ephemeral storage for caches

## 0.10.0 / 2017-06-18

Secure connection support! This version adds support for SSL certificates.

### Short list of commit messages

  * Update docker-entrypoint to import PKCS12 certificate found in JIRA_HOME directory at startup.
  * Add sysadmin login troubleshooting info when migrating from Cloud to Server.

## 0.9.0 / 2017-06-11

Initial release! A _dockerized_ [Atlassian Jira](https://www.atlassian.com/software/jira) install.

### Short list of commit messages

  * Update README for v0.9.0.


docker-entrypoint.sh
#!/bin/bash

# check if the `server.xml` file has been changed since the creation of this
# Docker image. If the file has been changed the entrypoint script will not
# perform modifications to the configuration file.
if [ "$(stat --format "%Y" "${JIRA_INSTALL}/conf/server.xml")" -eq "0" ]; then
  if [ -n "${X_PROXY_NAME}" ]; then
    xmlstarlet ed --inplace --pf --ps --insert '//Connector[@port="8080"]' --type "attr" --name "proxyName" --value "${X_PROXY_NAME}" "${JIRA_INSTALL}/conf/server.xml"
  fi
  if [ -n "${X_PROXY_PORT}" ]; then
    xmlstarlet ed --inplace --pf --ps --insert '//Connector[@port="8080"]' --type "attr" --name "proxyPort" --value "${X_PROXY_PORT}" "${JIRA_INSTALL}/conf/server.xml"
  fi
  if [ -n "${X_PROXY_SCHEME}" ]; then
    xmlstarlet ed --inplace --pf --ps --insert '//Connector[@port="8080"]' --type "attr" --name "scheme" --value "${X_PROXY_SCHEME}" "${JIRA_INSTALL}/conf/server.xml"
  fi
  if [ -n "${X_PROXY_SECURE}" ]; then
    xmlstarlet ed --inplace --pf --ps --insert '//Connector[@port="8080"]' --type "attr" --name "secure" --value "${X_PROXY_SECURE}" "${JIRA_INSTALL}/conf/server.xml"
  fi
  if [ -n "${X_PATH}" ]; then
    xmlstarlet ed --inplace --pf --ps --update '//Context/@path' --value "${X_PATH}" "${JIRA_INSTALL}/conf/server.xml"
  fi
fi

if [ -f "${CERTIFICATE}" ] || [ -f "${CERTIFICATE}.p12" ]; then
  # convert PKCS12 certificate format to JKS certificate format
  #
  # To generate a pkcs12 file from an openssl self-signed cert and key file:
  #   > openssl pkcs12 -export -in server_cert.pem -inkey server_key.pem -out certificate.p12
  #       -passout pass:changeit -name "jira"
  #
  # To test the insertion:
  #   > docker exec -it <CONTAINER_ID> /bin/bash
  #   > keytool -list -keystore $JAVA_HOME/jre/lib/security/cacerts -v | grep Alias | grep jira
  #
  if [[ "${CERTIFICATE}" =~ .p12$ || -f "${CERTIFICATE}.p12" ]]; then
    keytool -noprompt -storepass changeit -importkeystore \
      -srckeystore ${CERTIFICATE%.p12}.p12 -srcstoretype PKCS12 -srcstorepass changeit -alias jira \
      -destkeystore ${JAVA_CACERTS} -deststoretype JKS -deststorepass changeit -destalias jira
  else
    keytool -noprompt -storepass changeit \
      -keystore ${JAVA_CACERTS} -import -file ${CERTIFICATE} -alias jira
  fi

  # Update the server.xml file
  # <!--
  # <Connector port="8443" protocol="org.apache.coyote.http11.Http11Protocol"
  #             maxHttpHeaderSize="8192" SSLEnabled="true"
  #             maxThreads="150" minSpareThreads="25"
  #             enableLookups="false" disableUploadTimeout="true"
  #             acceptCount="100" scheme="https" secure="true"
  #             clientAuth="false" sslProtocol="TLS" useBodyEncodingForURI="true"
  #             keyAlias="jira" keystoreFile="<JIRA_HOME>/jira.jks" keystorePass="changeit" keystoreType="JKS"/>
  # -->
  xmlstarlet ed --inplace --pf --ps \
    --subnode "Server/Service" --type elem --name "ConnectorTMP" --value "" \
    --insert  "//ConnectorTMP" --type attr --name "port" --value "8443" \
    --insert  "//ConnectorTMP" --type attr --name "protocol" --value "org.apache.coyote.http11.Http11NioProtocol" \
    --insert  "//ConnectorTMP" --type attr --name "maxHttpHeaderSize" --value "8192" \
    --insert  "//ConnectorTMP" --type attr --name "SSLEnabled" --value "true"  \
    --insert  "//ConnectorTMP" --type attr --name "maxThreads" --value "150" \
    --insert  "//ConnectorTMP" --type attr --name "minSpareThreads" --value "25" \
    --insert  "//ConnectorTMP" --type attr --name "enableLookups" --value "false" \
    --insert  "//ConnectorTMP" --type attr --name "disableUploadTimeout" --value "true" \
    --insert  "//ConnectorTMP" --type attr --name "acceptCount" --value "100" \
    --insert  "//ConnectorTMP" --type attr --name "scheme" --value "https" \
    --insert  "//ConnectorTMP" --type attr --name "secure" --value "true" \
    --insert  "//ConnectorTMP" --type attr --name "clientAuth" --value "false" \
    --insert  "//ConnectorTMP" --type attr --name "sslProtocol" --value "TLSv1.2" \
    --insert  "//ConnectorTMP" --type attr --name "sslEnabledProtocols" --value "TLSv1.2" \
    --insert  "//ConnectorTMP" --type attr --name "useBodyEncodingForURI" --value "true" \
    --insert  "//ConnectorTMP" --type attr --name "keyAlias" --value "jira" \
    --insert  "//ConnectorTMP" --type attr --name "keystoreFile" --value "${JAVA_CACERTS}" \
    --insert  "//ConnectorTMP" --type attr --name "keystorePass" --value "changeit" \
    --insert  "//ConnectorTMP" --type attr --name "keystoreType" --value "JKS" \
    --rename  "//ConnectorTMP" --value "Connector" \
    "${JIRA_INSTALL}/conf/server.xml"

  # @TODO: Update Base URL to HTTPS
  #
  # https://confluence.atlassian.com/jirakb/how-do-i-manually-change-the-base-url-733940375.html

  # @TODO: Use xmlstarlet to update the web.xml file
  #
  # This will redirect all traffic to use HTTPS urls.
  #
  # Location: ${JIRA_INSTALL}/atlassian-jira/WEB-INF/web.xml
  # <!--
  # <security-constraint>
  #   <web-resource-collection>
  #     <web-resource-name>all-except-attachments</web-resource-name>
  #     <url-pattern>*.jsp</url-pattern>
  #     <url-pattern>*.jspa</url-pattern>
  #     <url-pattern>/browse/*</url-pattern>
  #     <url-pattern>/issues/*</url-pattern>
  #   </web-resource-collection>
  #   <user-data-constraint>
  #     <transport-guarantee>CONFIDENTIAL</transport-guarantee>
  #   </user-data-constraint>
  # </security-constraint>
  # -->
fi

exec "$@"
